File: OOP/inheritance.py
class Grandpa:
    def __init__(self, grandpa_name):
        self.grandfather_name = grandpa_name


class Father(Grandpa):
    def __init__(self, father_name, grandpa_name):
        self.father_name = father_name
        super().__init__(grandpa_name)


class Son(Father):
    def __init__(self, name, father_name, grandfather_name):
        self.name = name
        super().__init__(father_name, grandfather_name)

    def show_details(self):
        print(f"I am {self.name}, son of Mr. {
              self.father_name}. My grandfather's name is: {self.grandfather_name}")


good_son = Son("Mohammad", "Kamal", "Jamal")
good_son.show_details()

File: OOP/inheritance.cpp
#include <bits/stdc++.h>

using namespace std;

class Grandpa
{
public:
    Grandpa(string grandpa_name) : grandfather_name(grandpa_name) {}

    string grandfather_name;
};

class Father : public Grandpa
{
public:
    Father(string father_name, string grandpa_name)
        : Grandpa(grandpa_name), father_name(father_name) {}

    string father_name;
};

class Son : public Father
{
public:
    Son(string name, string father_name, string grandfather_name)
        : Father(father_name, grandfather_name), name(name) {}

    void show_details()
    {
        cout << "I am " << name << ", son of Mr. " << father_name
             << ". My grandfather's name is: " << grandfather_name << endl;
    }

private:
    string name;
};

int main()
{
    Son good_son("Mohammad", "Kamal", "Jamal");
    good_son.show_details();

    return 0;
}
